base_layer/p2p/src/services/liveness/service.rs

```diff
@@ -0,0 +1,223 @@
+pub struct LivenessService {
+    state: Arc<LivenessState>,
+    handles: Arc<ServiceHandles>,
+    oms_handle: Option<CommsOutboundHandle>,
+}
+
+impl LivenessService {
+    pub fn new(state: Arc<LivenessState>, handles: Arc<ServiceHandles>) -> Self {
+        Self {
+            state,
+            handles,
+            oms_handle: None,
+        }
+    }
+
+    /// Get the CommsOutboundHandle.
+    ///
+    /// This function will panic if the handle does not exist.
+    fn oms_handle(&mut self) -> &mut CommsOutboundHandle {
+        match self.oms_handle {
+            Some(ref mut h) => h,
+            None => {
+                // Memoize the handle for later use
+                self.oms_handle = Some(
+                    self.handles
+                        .get_handle::<CommsOutboundHandle>(&ServiceName::CommsOutbound)
+                        .expect("LivenessService requires CommsOutboundHandle"),
+                );
+                self.oms_handle()
+            },
+        }
+    }
+
+    fn send_ping(
+        &mut self,
+        pub_key: CommsPublicKey,
+    ) -> impl Future<Item = Result<LivenessResponse, LivenessError>, Error = ()>
+    {
+        let state = self.state.clone();
+        self.oms_handle()
+            .send_message(
+                BroadcastStrategy::DirectPublicKey(pub_key),
+                MessageFlags::empty(),
+                TariMessageType::new(NetMessage::PingPong),
+                PingPong::Ping,
+            )
+            .and_then(move |res| {
+                state.inc_pings_sent();
+                future::ok(
+                    res.map(|_| LivenessResponse::PingSent)
+                        .map_err(LivenessError::CommsOutboundError),
+                )
+            })
+            .or_else(|_| future::ok(Err(LivenessError::SendPingFailed)))
+    }
+
+    fn get_ping_count(&self) -> usize {
+        self.state.pings_received()
+    }
+
+    fn get_pong_count(&self) -> usize {
+        self.state.pongs_received()
+    }
+}
+
+impl Service<LivenessRequest> for LivenessService {
+    type Error = ();
+    type Response = Result<LivenessResponse, LivenessError>;
+
+    existential type Future: Future<Item = Self::Response, Error = Self::Error>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        Ok(().into())
+    }
+
+    fn call(&mut self, req: LivenessRequest) -> Self::Future {
+        match req {
+            LivenessRequest::SendPing(pub_key) => Either::A(
+                self.send_ping(pub_key)
+                    .or_else(|_| future::ok(Err(LivenessError::SendPingFailed))),
+            ),
+            LivenessRequest::GetPingCount => Either::B(future::ok(Result::<_, LivenessError>::Ok(
+                LivenessResponse::Count(self.get_ping_count()),
+            ))),
+            LivenessRequest::GetPongCount => Either::B(future::ok(Result::<_, LivenessError>::Ok(
+                LivenessResponse::Count(self.get_pong_count()),
```

Comment:
he he. Sneaky

Reply:
hehe yup - not quite sure how to deal with the case where there is a cheap synchronous operation to be done - creating a blocking future seems excessive. In this case, AtomicUsize with Relaxed (monotonic) atomicity is appropriate, so there isn't even a lock to contend with.